### PR TITLE
Fix #20, adds constructor given just maximum length

### LINKED ISSUE #20: A function that returns the right constructor given a max length
I would like a way to simplify this code.

```julia
function shorten_strings(vec::AbstractVector{<:AbstractString})
    isempty(vec) && return vec

    # chose a type big enough for everything, to avoid large union
    max_len = maximum(ncodeunits, skipmissing(vec))
    return if max_len <= 7
        map(ShortString7, vec)
    elseif max_len <= 15
        map(ShortString15, vec)
    elseif max_len <= 30
        map(ShortString30, vec)
    elseif max_len <= 62
        map(ShortString62, vec)
    elseif max_len <= 126
        map(ShortString126, vec)
    else
        vec
    end
end
```

or even the more complex version that support missings:
```julia
using Missings: passmissing

function shorten_strings(vec::AbstractVector{<:Union{Missing, AbstractString}})
    isempty(vec) && return vec

    # chose a type big enough for everything, to avoid large union
    max_len = maximum(ncodeunits, skipmissing(vec))
    return if max_len <= 7
        map(passmissing(ShortString7), vec)
    elseif max_len <= 15
        map(passmissing(ShortString15), vec)
    elseif max_len <= 30
        map(passmissing(ShortString30), vec)
    elseif max_len <= 62
        map(passmissing(ShortString62), vec)
    elseif max_len <= 126
        map(passmissing(ShortString126), vec)
    else
        vec
    end
end
```

I am not sure what the best solution is.

### PR DESCRIPTION
**Pull request by ScottPJones**
This also improves `isvalid(str, pos)` to not have to convert the `ShortString` to a `String`.
It adds docstrings to most all functions, including internal ones.
It updates the README file (the README.jmd was updated, and the build-readme script was run)

### CODE REVIEW

File: src/base.jl [hunk 1]
```diff
@@ -221,7 +236,10 @@ size_content(s::ShortString) = s.size_content
 
 @define_integers 2048 Int2048 UInt2048
 
-for T in (UInt2048, UInt1024, UInt512, UInt256, UInt128, UInt64, UInt32)
+"""These are the default types used to for selecting the size of a ShortString"""
+const def_types = [UInt32, UInt64, UInt128, UInt256, UInt512, UInt1024, UInt2048]
```
Comment (oxinabox) [inline src/base.jl:240]: Why not a tuple? it is immutable
```suggestion
const def_types = (UInt32, UInt64, UInt128, UInt256, UInt512, UInt1024, UInt2048)
```

File: src/base.jl [hunk 2]
```diff
@@ -232,6 +250,26 @@ for T in (UInt2048, UInt1024, UInt512, UInt256, UInt128, UInt64, UInt32)
     end
 end
 
+"""
+Return a ShortString type that can hold maxlen codeunits
```
Comment (oxinabox) [inline src/base.jl:254]:
```suggestion
Return a ShortString type (constructor) that can hold at least maxlen codeunits
```

File: src/base.jl [hunk 3]
```diff
@@ -232,6 +250,26 @@ for T in (UInt2048, UInt1024, UInt512, UInt256, UInt128, UInt64, UInt32)
...
+"""
+function get_type(maxlen; types=def_types)
+    for T in types
+        maxlen <= sizeof(T) - size_bytes(T) && return ShortString{T}
+    end
+    throw(ArgumentError("$maxlen is too large to fit into any of the provided types: $types"))
+end
+
+ShortString(str::Union{String,SubString{String}}, maxlen = 0; types=def_types) =
+    get_type(maxlen <= 0 ? sizeof(str) : maxlen, types=types)(str)
```
Comment (oxinabox) [inline src/base.jl:267]:
```suggestion
ShortString(str::Union{String,SubString{String}}, maxlen = sizeof(str); types=def_types) =
    get_type(maxlen, types=types)(str)
```


If the user provides a maxlen of 0 or less it should error.

File: src/base.jl [hunk 4]
```diff
@@ -232,6 +250,26 @@ for T in (UInt2048, UInt1024, UInt512, UInt256, UInt128, UInt64, UInt32)
...
+
+macro ss_str(str, max="0")
```
Comment (oxinabox) [inline src/base.jl:269]: Needs a docstring, and a compat note?
Do we want to add the ability to accept `x7` etc for compat with older julia version?

**Review by oxinabox — APPROVED** [review]
Needs tests, especially of `isvalid`

File: src/base.jl (hunk 4 above)
Comment (ScottPJones) [reply src/base.jl:269]: The other string macros haven't been removed (although I'd like to).
Same thing with all of the type alias `ShortString3`...`ShortString255`.
`ShortString(str, 255)` I think is just as easy to use.

Comment (ScottPJones) [reply src/base.jl:269]: I'll add the docstring and the compat note in the next round, OK?

**Pull request merged**